### PR TITLE
JetBrains: Add storybook for Chromatic tests

### PR DESCRIPTION
**Pull request by vdavid**
Fixes https://github.com/sourcegraph/sourcegraph/issues/37681

I made webpack work with the project (needed some polyfills) and added a story.

Also:
 - Bug fix: The initial search was broken in the standalone version. Now it correctly triggers.
 - Bug fix: Path matches didn't send the "commit" field so the "Copy URL" feature didn't work for remote files in path matches. (I've found this out by accident during testing)
 - Extracted the mocked "callJava" function: it's reusable now (I've reused it) and more elegant too.

This is how it looks in both modes:

<div>
CleanShot 2022-07-12 at 19 49 45@2x
CleanShot 2022-07-12 at 19 52 11@2x
<div>

Caveat:
 - To avoid a flaky test and make sure that the search finishes by the time Chromatic creates its snapshot, I set up a 10 second delay after page loading. I imagine that Chromatic works in a heavily multithreaded way (as we're generating hundreds of pages) so this should be non-blocking. I've tried to find a more dynamic solution, but haven't found anything in the docs and on Google, so I stayed with the fixed delay.

## Test plan

I've tested all of these:
- It works with `yarn storybook:jetbrains`
- It works with `yarn storybook`
- It works with both dark and light modes correctly
- It still works with the standalone version
- It still works in the IDE

## App preview:

- [Web](https://sg-web-dv-jetbrains-chromatic.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-tmnegtqwet.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by vdavid** [comment]
@taras-yemets @philipp-spiess any ideas how to make the CSS work?
(Can be tested by checking out the branch and running `yarn storybook` in the sourcegraph root.)

**Comment by philipp-spiess** [comment]
@vdavid I was so free and pushed something onto your branch that should help with that. We have to manually add the global stylesheets. Take a look at https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph.git&branch=HEAD&file=client%2Fbranded%2Fsrc%2Fcomponents%2FBrandedStory.tsx&editor=VSCode&version=2.2.7&start_row=0&start_col=12&end_row=0&end_col=12&utm_campaign=vscode-extension&utm_medium=direct_traffic&utm_source=vscode-extension&utm_content=vsce-commands where we do this for other storybook tests.

I also added the global `theme` and `theme-light` config but I think we ideally want to run [this code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/jetbrains/webview/src/search/index.tsx?L95-129&subtree=true) with one of the theme snapshots to also have the custom CSS vars defined.

File: client/jetbrains/package.json [hunk 1]
```diff
@@ -13,13 +13,14 @@
...
-    "build": "yarn task:gulp esbuild ",
-    "watch": "WATCH=true yarn task:gulp esbuild",
-    "standalone": "ts-node standalone/src/server.ts",
+    "build": "yarn task:gulp esbuild",
     "lint": "yarn run lint:js && yarn run lint:css",
     "lint:js": "eslint --cache 'webview/**/*.[jt]s?(x)'",
     "lint:css": "stylelint 'webview/**/*.scss'",
-    "typecheck": "tsc -b"
+    "standalone": "ts-node standalone/src/server.ts",
+    "storybook": "STORIES_GLOB=client/jetbrains/webview/src/**/*.story.tsx yarn workspace @sourcegraph/storybook run start",
```
Comment (philipp-spiess) [inline client/jetbrains/package.json:21]: The re-ordering makes it really hard to see what the change is, Is it only the addition of this command? If so then LGTM

**Review by philipp-spiess — COMMENTED** [review]
> I imagine that Chromatic works in a heavily multithreaded way (as we're generating hundreds of pages) so this should be non-blocking. 

Any way we can be sure about this? How do other snapshots handle this? And even if these tests are parallelized, it could still be part of the slowest path in the graph and thus delay our whole run by 10 seconds. 

I also noticed that we currently do not apply the JetBrains specific CSS vars. Do you think it is okay to not include them or should we add them based on the theming snapshots that we already have bundled? 

Also: Does it make sense to include other result types in this static snapshot so that we have a higher coverage?

**Review by philipp-spiess — APPROVED** [review]
Stamping it regardless so you can work out answers to my above questions and ship what you find useful without having to wait for another review :)

File: client/jetbrains/webview/src/bridge-mock/call-java-mock.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { decode } from 'js-base64'
 
-import { SearchPatternType } from '@sourcegraph/shared/out/src/graphql-operations'
+import { SearchPatternType } from '@sourcegraph/search'
```
Comment (vdavid) [inline client/jetbrains/webview/src/bridge-mock/call-java-mock.ts:3]: Ah, thanks, @philipp-spiess, apparently that was me back in May.

Comment (philipp-spiess) [reply client/jetbrains/webview/src/bridge-mock/call-java-mock.ts:3]: Haha yeah the build was failing because of this 😬  No worries!

**Comment by vdavid** [comment]
> > I imagine that Chromatic works in a heavily multithreaded way (as we're generating hundreds of pages) so this should be non-blocking.
> Any way we can be sure about this? How do other snapshots handle this? And even if these tests are parallelized, it could still be part of the slowest path in the graph and thus delay our whole run by 10 seconds.

I tried looking for examples, but we do no network requests in any of our storybook pages where Chromatic is enabled.
So I removed the App test, and added two individual tests, one for the search box, one for the result list.

> I also noticed that we currently do not apply the JetBrains specific CSS vars. Do you think it is okay to not include them or should we add them based on the theming snapshots that we already have bundled?

I saw you added this yesterday, thanks for that!

> Also: Does it make sense to include other result types in this static snapshot so that we have a higher coverage?

It does. I've added them to the new tests. This is how they look:

CleanShot 2022-07-15 at 11 27 29@2x

CleanShot 2022-07-15 at 11 27 43@2x

@philipp-spiess WDYT?

File: client/jetbrains/webview/src/bridge-mock/call-java-mock.ts (hunk 2 above)
Comment (vdavid) [reply client/jetbrains/webview/src/bridge-mock/call-java-mock.ts:3]: I'm not sure how that's possible. It's a code line that I moved here from `bridge-mock/index.ts`.

See? It's in `main` now: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/jetbrains/webview/src/bridge-mock/index.ts?L3%3A25=

**Comment by philipp-spiess** [comment]
@vdavid 🚀  Awesome!!

**Pull request merged**